File: src/main/java/dreamdiary/quiz/infra/QuizEntity.java
package dreamdiary.quiz.infra;

import dreamdiary.quiz.domain.model.Quiz;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import java.time.LocalDateTime;
import java.util.List;

@Builder
@EntityListeners(AuditingEntityListener.class)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "QUIZ", indexes = {
        @Index(name = "IDX_QUIZ_PUBLIC_ID", columnList = "public_id")
})
@Entity
@Getter
class QuizEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Column(name = "public_id", nullable = false, unique = true)
    private String publicId;
    @Column(name = "title", nullable = false, unique = true)
    private String title;
    @Column(name = "content", nullable = false)
    private String content;
    @Column(name = "release_at", nullable = false)
    private LocalDateTime releaseAt;
    @Column(name = "answer_release_at", nullable = false)
    private LocalDateTime answerReleaseAt;
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "quiz_id")
    private List<ChoiceEntity> choices;
    @CreatedDate
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;
    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public static QuizEntity mapped(final Quiz quiz) {
        final List<ChoiceEntity> choiceEntities = quiz.getChoices().values().stream()
                .map(ChoiceEntity::mapped)
                .toList();
        return builder()
                .publicId(quiz.getQuizPublicId().value())
                .title(quiz.getTitle().value())
                .content(quiz.getContent().value())
                .choices(choiceEntities)
                .releaseAt(quiz.getReleaseAt())
                .answerReleaseAt(quiz.getAnswerReleaseAt())
                .build();
    }
}

File: applications/quiz/src/main/kotlin/dreamdiary/nineoclock/quiz/infrastructure/QuizEntity.kt
package dreamdiary.nineoclock.quiz.infrastructure

import dreamdiary.nineoclock.quiz.domain.model.Choice
import dreamdiary.nineoclock.quiz.domain.model.ChoiceGroup
import dreamdiary.nineoclock.quiz.domain.model.Quiz
import dreamdiary.nineoclock.quiz.domain.model.QuizContent
import dreamdiary.nineoclock.quiz.domain.model.QuizTitle
import dreamdiary.nineoclock.shard.identifier.QuizId
import dreamdiary.nineoclock.shard.identifier.QuizPublicId
import dreamdiary.nineoclock.shard.identifier.QuizSecureId
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.hibernate.annotations.BatchSize
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

@EntityListeners(AuditingEntityListener::class)
@BatchSize(size = 10)
@Table(
    name = "quiz",
    indexes = [
        Index(name = "IDX_QUIZ_PUBLIC_ID", columnList = "public_id"),
        Index(name = "IDX_QUIZ_SECURE_ID", columnList = "secure_id"),
    ]
)
@Entity
internal class QuizEntity(
    @Column(name = "public_id", nullable = false, unique = true)
    val publicId: String,
    @Column(name = "secure_id", nullable = false, unique = true)
    val secureId: String,
    @Column(name = "title")
    val title: String,
    @Column(name = "content")
    val content: String,
    @OneToMany(cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    @JoinColumn(name = "quiz_id")
    val choices: List<ChoiceEntity> = mutableListOf(),
    @Column(name = "release_at")
    val releaseAt: LocalDateTime,
    @Column(name = "answer_release_at")
    val answerReleaseAt: LocalDateTime,
    @CreatedDate @Column(name = "created_at", nullable = false)
    var createdAt: LocalDateTime? = null,
    @LastModifiedDate @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime? = null,
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null
)

internal fun Quiz.toEntity(): QuizEntity {
    return QuizEntity(
        publicId = this.id.publicId.value,
        secureId = this.id.secureId.value,
        title = this.title.value,
        content = this.content.value,
        choices = this.choiceGroup.values.map(Choice::toEntity),
        releaseAt = this.releaseAt,
        answerReleaseAt = this.answerReleaseAt
    )
}

internal fun QuizEntity.toModel() : Quiz {
    return Quiz(
        title = QuizTitle(this.title),
        content = QuizContent(this.content),
        choiceGroup = ChoiceGroup(
            values = this.choices.map { it.toModel() }
        ),
        releaseAt = this.releaseAt,
        answerReleaseAt = this.answerReleaseAt,
        id = QuizId(
            publicId = QuizPublicId(this.publicId),
            secureId = QuizSecureId(this.secureId)
        )
    )
}
